### ISSUE TITLE
Adaptive filmstrip interval for extremely long clips

### ISSUE THREAD

**Issue by Orthodox-531** [whyboris/Video-Hub-App#550]
Since the default thumbnail size is 288*512 pixels, an image snapshot interval of 1 minute for a clip of 128 minutes will be 65536 pixels wide, throwing an error in FFMPEG because the maximum width of a JPEG file is 65500 pixels.

One option I can think of to work around this issue is to add some adaptive math to the interval calculation to adjust the interval to a higher value where necessary to keep the output file to 65500 pixels or less.

Alternatively, PNG could be used, which does not have a maximum dimension specified.

**Comment by whyboris** [comment]
Thank you for raising this issue. I'm aware of the jpg limitation and it's unfortunate I don't have a solution for it yet.

`png` is much more space-consuming, so I'd prefer not to use them.

I think your solution is the best we have at least for now: since at time of generating screenshots we know all the numbers involved, we could arbitrarily reduce the number of screens that should be extracted to the maximum that will fit inside the `jpg` limit 😓

**Comment by whyboris** [comment]
As a brief overview:

The screenshot generation happens via `main-extract.ts` file with this method generating the arguments for `ffmpeg`:
https://github.com/whyboris/Video-Hub-App/blob/main/node/main-extract.ts#L83
It gets called from this line:
https://github.com/whyboris/Video-Hub-App/blob/main/node/main-extract.ts#L308
So we need to make sure the `numOfScreens` in the `imageElement` doesn't max things out:
https://github.com/whyboris/Video-Hub-App/blob/main/node/main-extract.ts#L256
We could handle it as a check inside the `extractAll` method:
https://github.com/whyboris/Video-Hub-App/blob/main/node/main-extract.ts#L238

But it's almost surely better to do before then, when we are setting the number of screenshots originally:
https://github.com/whyboris/Video-Hub-App/blob/main/node/main-support.ts#L362
When we `extractMetadataAsync` we call `computeNumberOfScreenshots` which sets the number:
https://github.com/whyboris/Video-Hub-App/blob/main/node/main-support.ts#L387
And we just need to mess with just this method:
https://github.com/whyboris/Video-Hub-App/blob/main/node/main-support.ts#L296

🚀

**Comment by whyboris** [comment]
TL;DR:
When the `computeNumberOfScreenshots` is called here:
https://github.com/whyboris/Video-Hub-App/blob/main/node/main-support.ts#L294
We just need to use the width from the `screenshotSettings` argument to determine if the number of screenshots we need to extract is fewer than otherwise would happen 👌 

Should be a quick fix 🤷‍♂️

**Comment by Orthodox-531** [comment]
Sadly, my realm is .Net and PowerShell, but I was able to understand enough to find the root cause... but I wouldn't be able to write enough Javascript to give you a PR.

**Comment by whyboris** [comment]
No worries -- I just fixed it 😉 with #554 and a few lines of code: https://github.com/whyboris/Video-Hub-App/pull/554/commits/569750bc6f04d3292b8724e1296c81161196ace7 😎

**Comment by whyboris** [comment]
Thank you @Orthodox-531 for the great catch 🙇 this will surely prevent some empty filmstrips (or even errors during extraction!) 😁

**Comment by Orthodox-531** [comment]
So it seems that the fix addresses the issue for new clips being added, but in cases where a user is upgrading from 2.x, any clips which were already in the library which had saved metadata for too many screens will continue to be impacted by the original issue.

Perhaps it would be possible to do a quick check of the library for items which have total * screenWidth > 65535 during the database upgrade, and force the floor value at that time?

**Comment by whyboris** [comment]
Correct -- the current code will only affect newly-extracted screenshots.

Thank you for the good idea! It seems relatively straightforward to do with the code I have, just call the `computeNumberOfScreenshots` method within the `upgradeToVersion3` method - for each element in the image array 🤔 

I'll test it out first, if it works easily, I'll add this improvement to the codebase 🤝 

ps - if you'd like to try your hand at it, let me know 👍

**Comment by whyboris** [comment]
Thank you @Orthodox-531 -- I added a fix: #592 -- ended up being just 1 line of code 😁
